src/i2p.cpp

```diff
@@ -0,0 +1,406 @@
+
+    if (!sock) {
+        throw std::runtime_error("Cannot create socket");
+    }
+
+    if (!ConnectSocketDirectly(m_control_host, sock->Get(), nConnectTimeout, true)) {
+        throw std::runtime_error(strprintf("Cannot connect to %s", m_control_host.ToString()));
+    }
+
+    SendRequestAndGetReply(*sock, "HELLO VERSION MIN=3.1 MAX=3.1");
+
+    return std::move(*sock);
+}
+
+void Session::CheckControlSock()
+{
+    LOCK(m_mutex);
+
+    std::string errmsg;
+    if (!m_control_sock.IsConnected(errmsg)) {
+        Log("Control socket error: %s", errmsg);
+        Disconnect();
+    }
+}
+
+void Session::DestGenerate(const Sock& sock)
+{
+    // https://geti2p.net/spec/common-structures#key-certificates
+    // "7" or "EdDSA_SHA512_Ed25519" - "Recent Router Identities and Destinations".
+    // Use "7" because i2pd <2.24.0 does not recognize the textual form.
+    const Reply& reply = SendRequestAndGetReply(sock, "DEST GENERATE SIGNATURE_TYPE=7", false);
+
+    m_private_key = DecodeI2PBase64(reply.Get("PRIV"));
+}
+
+void Session::GenerateAndSavePrivateKey(const Sock& sock)
+{
+    DestGenerate(sock);
+
+    // umask is set to 077 in init.cpp, which is ok (unless -sysperms is given)
+    if (!WriteBinaryFile(m_private_key_file,
+                         std::string(m_private_key.begin(), m_private_key.end()))) {
+        throw std::runtime_error(
+            strprintf("Cannot save I2P private key to %s", m_private_key_file));
+    }
+}
+
+Binary Session::MyDestination() const
+{
+    // From https://geti2p.net/spec/common-structures#destination:
+    // "They are 387 bytes plus the certificate length specified at bytes 385-386, which may be
+    // non-zero"
+    static constexpr size_t DEST_LEN_BASE = 387;
+    static constexpr size_t CERT_LEN_POS = 385;
+
+    uint16_t cert_len;
+    memcpy(&cert_len, &m_private_key.at(CERT_LEN_POS), sizeof(cert_len));
+    cert_len = be16toh(cert_len);
+
+    const size_t dest_len = DEST_LEN_BASE + cert_len;
+
+    return Binary{m_private_key.begin(), m_private_key.begin() + dest_len};
+}
+
+void Session::CreateIfNotCreatedAlready()
+{
+    std::string errmsg;
+    if (m_control_sock.IsConnected(errmsg)) {
+        return;
+    }
+
+    Log("Creating SAM session with %s", m_control_host.ToString());
+
+    Sock sock = Hello();
+
+    const auto& [read_ok, data] = ReadBinaryFile(m_private_key_file);
+    if (read_ok) {
+        m_private_key.assign(data.begin(), data.end());
+    } else {
+        GenerateAndSavePrivateKey(sock);
+    }
+
+    const std::string& session_id = GetRandHash().GetHex().substr(0, 10); // full is an overkill, too verbose in the logs
+    const std::string& private_key_b64 = SwapBase64(EncodeBase64(m_private_key));
+
+    SendRequestAndGetReply(sock, strprintf("SESSION CREATE STYLE=STREAM ID=%s DESTINATION=%s",
+                                           session_id, private_key_b64));
+
+    m_my_addr = CService(DestBinToAddr(MyDestination()), Params().GetDefaultPort());
```

Comment:
```suggestion
    m_my_addr = CService{DestBinToAddr(MyDestination()), static_cast<uint16_t>(Params().GetDefaultPort())};
```

Reply:
Why?

Reply:
To be explicit about the narrowing conversion, same as https://github.com/bitcoin/bitcoin/pull/20685#discussion_r583893589 above

Reply:
Being explicit with a static_cast has no benefits I'm aware of, just would silence warnings in cases where the warning is warranted...?

Reply:
@luke-jr I think it's the opposite, brace (uniform) initialization raises compile time warnings, e.g. for narrowing and conversions, and named casts provide compile time checks as well.

- https://google.github.io/styleguide/cppguide.html#Casting
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named

"Use brace initialization to convert arithmetic types (e.g., int64{x}). This is the safest approach because code will not compile if conversion can result in information loss. The syntax is also concise. Use static_cast as the equivalent of a C-style cast that does value conversion."

"Brace initialization makes it clear that the type conversion was intended and also prevents conversions between types that might result in loss of precision. (It is a compilation error to try to initialize a float from a double in this fashion, for example.)"

Reply:
For example, without the named cast here, brace initialization raises a compile time error for the narrowing conversion.

Reply:
The thing is that here `GetDefaultPort()` returns `int` and `CService` constructor takes `uint16_t` argument. So we would get a justified warning that we want to silence. The proper fix would be to change `GetDefaultPort()` to return `uint16_t` which is out of the scope of this PR.

I don't see a point in using a brace initialization for the warning and then to silence the warning with a `static_cast`.

```cpp
int x = 300;

// no warning
uint16_t c1 = x;

// non-constant-expression cannot be narrowed from type 'int' to 'uint16_t'
uint16_t c2{x};

// no warning
uint16_t c3{static_cast<uint16_t>(x)};
```

Reply:
I agree it can be considered out of scope to change `GetDefaultPort()`, but the suggested change would make explicit the type mismatch and maybe encourage improving it. As mentioned in the parent comment, none of these are blockers, just items I noted while reviewing.

Reply:
Anyway, no worries. I'll look into improving it.

Reply:
Proposed #21328 to address it without requiring any changes here.